{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nanopore DRS oversplitting detection\n",
    "\n",
    "We can attempt to detect where the MinKNOW software has oversplit reads by looking for reads which map adjacently that were also sequenced consequectively through the same pore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATEAAABQCAYAAABxukIjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAAu1JREFUeJzt20FqG2cYgOHPtUFKAoPbRRGRQT6AlgI3m+56h4LBYEg2PYXpFax9sGEu4exyhEAOEBsVE5E2GahajIS6yCrgDJLdYfwpz7OcXz98vxAvoxHaWi6XywBI6ru2BwC4DxEDUhMxIDURA1ITMSA1EQNSEzEgNREDUhMxIDURA1LbqVusqiqqqvri2s3NTVxdXcX+/n5sb283Ohzw7VgsFjGdTmM4HEa32115X23Ezs7OYjwe33s4gFWVZRmj0Wjl12/V/QH8tjuxyWQSR0dHcfnst5g/2r37pA/Yqz+ftz1Co3756UXbIzTmzc9v2x6hUe/Of217hMZ8mH2K31+/jIuLixgMBivvq70TK4oiiqK4dW3+aDfmj39Yb8okfpzN2x6hUfPdx22P0JinvdqPdHqzJ9+3PULj1n1M5cE+kJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJCaiAGpiRiQmogBqYkYkJqIAamJGJDaTt1iVVVRVdUX1yaTyeeN/3xsbqqWvf+39m1Jb+fjrO0RGvPH9bztERo1/fuvtkdozIfZp4iIWCwWa+3bWi6Xy68tnp6exng8vt9kAGsoyzJGo9HKr6+N2G13YpeXl3F8fBzn5+fR7/fvPukDdX19HYeHh1GWZfR6vbbH+d85X16bfLaIz3dg0+k0hsNhdLvdlffVfm8qiiKKorh1rd/vx97e3npTJtLr9ZwvsU0+3yafbTAYrL3Hg30gNREDUhMxILXtk5OTk3U3dTqdODg4iE6n08BI7XO+3Db5fJt8truq/XUS4KHzdRJITcSA1EQMSE3EgNREDEjtP9A0oPgEOj7wAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 360x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import sys\n",
    "import re\n",
    "import os\n",
    "from glob import glob\n",
    "from collections import Counter, namedtuple, defaultdict\n",
    "from operator import add\n",
    "from functools import reduce\n",
    "import itertools as it\n",
    "from operator import attrgetter\n",
    "import random\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats, signal\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib_venn as vn\n",
    "from matplotlib.colors import ListedColormap\n",
    "import seaborn as sns\n",
    "\n",
    "import pysam\n",
    "\n",
    "## Default plotting params\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set(font='Arial')\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "style = sns.axes_style('white')\n",
    "style.update(sns.axes_style('ticks'))\n",
    "style['xtick.major.size'] = 2\n",
    "style['ytick.major.size'] = 2\n",
    "sns.set(font_scale=2, style=style)\n",
    "pal = sns.color_palette(['#0072b2', '#d55e00', '#009e73', '#f0e442', '#cc79a7'])\n",
    "cmap = ListedColormap(pal.as_hex())\n",
    "sns.set_palette(pal)\n",
    "sns.palplot(pal)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_read_mapping_locs(bam_fn):\n",
    "    mapping = []\n",
    "    with pysam.AlignmentFile(bam_fn) as bam:\n",
    "        for aln in bam.fetch():\n",
    "            mapping.append([\n",
    "                aln.query_name,\n",
    "                aln.reference_name,\n",
    "                aln.reference_start,\n",
    "                aln.reference_end,\n",
    "                ['+', '-'][aln.is_reverse],\n",
    "            ])\n",
    "    return pd.DataFrame(mapping, columns=['read_id', 'chrom', 'genomic_start', 'genomic_end', 'strand'])\n",
    "\n",
    "\n",
    "def get_dist_to_prev(read):\n",
    "    if not read.strand_same_as_prev or not read.chrom_same_as_prev:\n",
    "        return np.nan\n",
    "    if read.strand == '+':\n",
    "        return read.genomic_start_prev_read - read.genomic_end\n",
    "    elif read.strand == '-':\n",
    "        return read.genomic_start - read.genomic_end_prev_read\n",
    "\n",
    "\n",
    "def read_sequencing_summary(ss_fn, bam_fn):\n",
    "    ss = []\n",
    "    for ss_fn in ss_fn:\n",
    "        ss.append(pd.read_csv(ss_fn, sep='\\t'))\n",
    "    ss = pd.concat(ss, axis=0)\n",
    "    ss = ss.sort_values(['channel', 'start_time'])\n",
    "    ss['end_time'] = ss['start_time'] + ss['duration']\n",
    "    ss['prev_read_id'] = ss.groupby('channel').read_id.shift(1)\n",
    "    ss['prev_end_time'] = ss.groupby('channel').end_time.shift(1)\n",
    "    ss['missing_signal_time'] = ss.start_time - ss.prev_end_time\n",
    "    mapped_loc = get_read_mapping_locs(bam_fn)\n",
    "    ss = ss.merge(mapped_loc, on='read_id', how='left')\n",
    "    ss = ss.merge(\n",
    "        mapped_loc,\n",
    "        left_on='prev_read_id',\n",
    "        right_on='read_id',\n",
    "        suffixes=('', '_prev_read'),\n",
    "        how='left'\n",
    "    ).drop('read_id_prev_read', axis=1)\n",
    "    ss['chrom_same_as_prev'] = ss.chrom == ss.chrom_prev_read\n",
    "    ss['strand_same_as_prev'] = ss.strand == ss.strand_prev_read\n",
    "    ss['genomic_dist_to_prev'] = ss.apply(get_dist_to_prev, axis=1)\n",
    "    return ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = [\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180201_1617_20180201_FAH45730_WT_Col0_2916_regular_seq/aligned_data/TAIR10/201901_col0_2916.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180227_1832_20180227_FAH59351_vir1_2922_DRS/aligned_data/TAIR10/20190317_vir1_2922.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180301_1825_20180301_FAH54216_VIRcomp_2928_DRS/aligned_data/TAIR10/20190317_VIRc_2928.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180413_1558_20180413_FAH77434_mRNA_WT_Col0_2917/aligned_data/TAIR10/201901_col0_2917.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180416_1534_20180415_FAH83697_mRNA_WT_Col0_2918/aligned_data/TAIR10/201901_col0_2918.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180418_1428_20180418_FAH83552_mRNA_WT_Col0_2919/aligned_data/TAIR10/201901_col0_2919.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180907_1420_20180907_VIRGFP_2929_mRNA/aligned_data/TAIR10/201902_VIRcompl_2929.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180912_1424_20180912_vir1_2923_mRNA/aligned_data/TAIR10/201902_vir1_2923.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180925_1320_20180925_VIRcompl_2931_mRNA/aligned_data/TAIR10/201812_VIRcompl_2931.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180928_1515_vir-1_20180929_2924_mRNA/aligned_data/TAIR10/201902_vir1_2924.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20181003_1345_20181003_mRNA_vir_2925/aligned_data/TAIR10/201902_vir1_2925.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20181019_1436_20181019_VIRcomp_mRNA_2930_Anya/aligned_data/TAIR10/201812_VIRcomp_2930.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180405_FAH59362_WT_Col0_2917/aligned_data/TAIR10/201903_col0_2917_exp2.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180411_1432_20180911_FAH84603_5adapterLIG_Col0_2916/aligned_data/TAIR10/201902_col0_2916_5adapter_exp2.bam',\n",
    "    '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180508_1522_20180508_FAH82422_5adapt_lig_mRNA_2918/aligned_data/TAIR10/201902_col0_2918_5adapter.bam'\n",
    "]\n",
    "\n",
    "\n",
    "sequencing_summaries= [\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180201_1617_20180201_FAH45730_WT_Col0_2916_regular_seq/rebasecalled/20180201_1617_20180201_FAH45730_WT_Col0_2916_regular_seq/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180201_1617_20180201_FAH45730_WT_Col0_2916_regular_seq/rebasecalled/20180201_1624_20180201_FAH45730_WT_Col0_2916_regular_seq/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180201_1617_20180201_FAH45730_WT_Col0_2916_regular_seq/rebasecalled/20180203_1507_20180203_FAH45730_WT_Col0_2916_regular_seq/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180201_1617_20180201_FAH45730_WT_Col0_2916_regular_seq/rebasecalled/20180203_1514_20180203_FAH45730_WT_Col0_2916_regular_seq/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180227_1832_20180227_FAH59351_vir1_2922_DRS/rebasecalled/20190317_vir1_2922/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_guppy_pipeline_runs/20180301_1825_20180301_FAH54216_VIRcomp_2928_DRS/rebasecalled/20190317_VIRc_2928/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180413_1558_20180413_FAH77434_mRNA_WT_Col0_2917/rebasecalled/20180413_1558_20180413_FAH77434_mRNA_WT_Col0_2917/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180413_1558_20180413_FAH77434_mRNA_WT_Col0_2917/rebasecalled/20180415_1454_20180415_FAH77434_mRNA_WT_Col0_2917_restart/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180413_1558_20180413_FAH77434_mRNA_WT_Col0_2917/rebasecalled/20180413_1605_20180413_FAH77434_mRNA_WT_Col0_2917/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180413_1558_20180413_FAH77434_mRNA_WT_Col0_2917/rebasecalled/20180415_1501_20180415_FAH77434_mRNA_WT_Col0_2917_restart/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180416_1534_20180415_FAH83697_mRNA_WT_Col0_2918/rebasecalled/20180416_1541_20180415_FAH83697_mRNA_WT_Col0_2918/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180416_1534_20180415_FAH83697_mRNA_WT_Col0_2918/rebasecalled/20180416_1534_20180415_FAH83697_mRNA_WT_Col0_2918/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180418_1428_20180418_FAH83552_mRNA_WT_Col0_2919/rebasecalled/20180418_1435_20180418_FAH83552_mRNA_WT_Col0_2919/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180418_1428_20180418_FAH83552_mRNA_WT_Col0_2919/rebasecalled/20180418_1428_20180418_FAH83552_mRNA_WT_Col0_2919/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180907_1420_20180907_VIRGFP_2929_mRNA/rebasecalled/20180907_1420_20180907_VIRGFP_2929_mRNA/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180912_1424_20180912_vir1_2923_mRNA/rebasecalled/20180912_1424_20180912_vir1_2923_mRNA/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180925_1320_20180925_VIRcompl_2931_mRNA/rebasecalled/20180925_1320_20180925_VIRcompl_2931_mRNA/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20180928_1515_vir-1_20180929_2924_mRNA/rebasecalled/20180928_1515_vir-1_20180929_2924_mRNA/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20181003_1345_20181003_mRNA_vir_2925/rebasecalled/20181003_1345_20181003_mRNA_vir_2925/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/ONT_pipeline_runs/20181019_1436_20181019_VIRcomp_mRNA_2930_Anya/rebasecalled/20181019_1436_20181019_VIRcomp_mRNA_2930_Anya/sequencing_summary.txt'],\n",
    "    ['/cluster/gjb_lab/nschurch/Projects/Arabidopsis_MinION_Datasets/20180405_col0_run010_tissue2917/albacore_2.2.7/rebasecalled/20180406_0934_20180406_2917_WT_restart_FAH59362/sequencing_summary.txt',\n",
    "     '/cluster/gjb_lab/nschurch/Projects/Arabidopsis_MinION_Datasets/20180405_col0_run010_tissue2917/albacore_2.2.7/rebasecalled/20180405_1615_20180405_FAH59362_WT_2917/sequencing_summary.txt',\n",
    "     '/cluster/gjb_lab/nschurch/Projects/Arabidopsis_MinION_Datasets/20180405_col0_run010_tissue2917/albacore_2.2.7/rebasecalled/20180406_0927_20180406_2917_WT_restart_FAH59362/sequencing_summary.txt',\n",
    "     '/cluster/gjb_lab/nschurch/Projects/Arabidopsis_MinION_Datasets/20180405_col0_run010_tissue2917/albacore_2.2.7/rebasecalled/20180405_1608_20180405_FAH59362_WT_2917/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/adapter_DRS_pipeline/201901_col0_2916/summaries/20180411_1432_20180911_FAH84603_5adapterLIG_Col0_2916/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/adapter_DRS_pipeline/201901_col0_2916/summaries/20180411_1439_20180911_FAH84603_5adapterLIG_Col0_2916/sequencing_summary.txt'],\n",
    "    ['/cluster/ggs_lab/mtparker/adapter_DRS_pipeline/201901_col0_2918/summaries/20180508_1529_20180508_FAH82422_5adapt_lig_mRNA_2918/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/adapter_DRS_pipeline/201901_col0_2918/summaries/20180510_1507_20180510_FAH82422_5adapt_lig_mRNA_2918_restart/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/adapter_DRS_pipeline/201901_col0_2918/summaries/20180510_1500_20180510_FAH82422_5adapt_lig_mRNA_2918_restart/sequencing_summary.txt',\n",
    "     '/cluster/ggs_lab/mtparker/adapter_DRS_pipeline/201901_col0_2918/summaries/20180508_1522_20180508_FAH82422_5adapt_lig_mRNA_2918/sequencing_summary.txt']\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cluster/ggs_lab/mtparker/.conda/envs/py3_sci/lib/python3.6/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  del sys.path[0]\n",
      "/cluster/ggs_lab/mtparker/.conda/envs/py3_sci/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3185: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  if (yield from self.run_code(code, result)):\n"
     ]
    }
   ],
   "source": [
    "oversplitting_res = {}\n",
    "oversplitting_stats = []\n",
    "for ss_fn, bam_fn in zip(sequencing_summaries, datasets):\n",
    "    sample_name = os.path.splitext(os.path.split(bam_fn)[1])[0]\n",
    "    ss = read_sequencing_summary(ss_fn, bam_fn)\n",
    "    ss_mapped_only = ss[~ss.chrom.isnull()]\n",
    "    is_oversplit = (\n",
    "        ss_mapped_only.chrom_same_as_prev & \n",
    "        ss_mapped_only.strand_same_as_prev & \n",
    "        (ss_mapped_only.genomic_dist_to_prev < 1000).fillna(False) &\n",
    "        (ss_mapped_only.genomic_dist_to_prev > -10).fillna(False)\n",
    "    )\n",
    "    ss_mapped_only['is_oversplit'] = is_oversplit\n",
    "    oversplitting_stats.append(\n",
    "        (sample_name,  is_oversplit.sum(), is_oversplit.sum() / len(is_oversplit) * 100)\n",
    "    )\n",
    "    oversplitting_res[sample_name] = ss_mapped_only\n",
    "\n",
    "oversplitting_stats = pd.DataFrame(oversplitting_stats, columns=['sample_name', 'total_oversplit', 'percent_oversplit'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>total_oversplit</th>\n",
       "      <th>percent_oversplit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>201901_col0_2916</td>\n",
       "      <td>473</td>\n",
       "      <td>0.046598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20190317_vir1_2922</td>\n",
       "      <td>459</td>\n",
       "      <td>0.044801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20190317_VIRc_2928</td>\n",
       "      <td>1238</td>\n",
       "      <td>0.110665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>201901_col0_2917</td>\n",
       "      <td>376</td>\n",
       "      <td>0.039915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>201901_col0_2918</td>\n",
       "      <td>148</td>\n",
       "      <td>0.019334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample_name  total_oversplit  percent_oversplit\n",
       "0    201901_col0_2916              473           0.046598\n",
       "1  20190317_vir1_2922              459           0.044801\n",
       "2  20190317_VIRc_2928             1238           0.110665\n",
       "3    201901_col0_2917              376           0.039915\n",
       "4    201901_col0_2918              148           0.019334"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oversplitting_stats.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3_sci]",
   "language": "python",
   "name": "conda-env-py3_sci-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
